deeplabcut/live/dlc_live.py

```diff
@@ -178,44 +234,79 @@ def setup_prediction(self, config):
             self.sess, self.inputs, self.outputs = predict.setup_pose_prediction(self.dlc_cfg)
 
     def get_pose(self, frame=None):
+        """
+
+        Parameters
+        ----------
+        frame :class:`numpy.ndarray`
+            image to extract pose. Should be an RGB, uint8 image, but if grayscale (len(frame.shape)== 2)
+            or if given as float will be automatically converted.
+
+        Returns
+        -------
+
+        """
 
         if frame is None:
-            if self.camera is None:
+            if self.frame_method is None:
                 raise DLCLiveException("No frame provided for live pose estimation")
             else:
-                frame = self.camera.frame
+                frame = self.frame_method()
 
-        frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
-        frame = img_as_ubyte(frame) if not self.cfg['cropping'] else img_as_ubyte(frame[self.cfg['y1']:self.cfg['y2'],self.cfg['x1']:self.cfg['x2']])
+        # These conversions can be pretty costly. Usually I like EAFP but afaik tensorflow doesn't handle
+        # it nicely, so let's
+        #   a) allow frame conversion to be handled by whatever is giving them to us
+        #   b) only do unambiguous operations - eg. grayscale to RGB, but how do we know if a (n,m,3) image is RGB or BGR?
+        #   c)  check if we need to 8-bit-ify it before doing so
+        # -jls 2020-03-09
+
+        if self.cfg['cropping']:
+            frame = frame[self.cfg['y1']:self.cfg['y2'],
+                          self.cfg['x1']:self.cfg['x2']]
+
+        # if we don't have a third axis, this is a grayscale image!!!
+        if len(frame.shape) == 2:
+            frame = cv2.cvtColor(frame, cv2.COLOR_GRAY2RGB)
+
+        if not frame.dtype == 'uint8':
+            frame = img_as_ubyte(frame)
+
+        # for each, expand dimensions to fit in with (batch_size, height, width, channels)
+        # expand_dims returns a view, so it's v cheap.
         if self.useFrozen:
-            pose = self.sess.run(self.outputs, feed_dict={self.inputs: np.expand_dims(frame, axis=0).astype(float)})
+            pose = self.sess.run(self.outputs, feed_dict={
+                self.inputs: np.expand_dims(frame, axis=0)
+            })
         elif self.TFGPUinference:
-            pose = self.sess.run(self.pose_tensor, feed_dict={self.inputs: np.expand_dims(frame, axis=0).astype(float)})
+            pose = self.sess.run(self.pose_tensor, feed_dict={
```

Comment:
please keep consistent with master DLC branch as much as possible; please revert

Reply:
no problem with reverting line formatting, but removing the recast to float speeds up the code pretty dramatically -- the model seemed to run fine without it, but you know it way better than i do obvs. in any case up to you :)

`.expand_dims` returns a numpy view which doesn't modify its representation in memory while `.astype(float)` will recast the array. 

Comparing `expand_dims(..)` with `expand_dims(...).astype(float)` with a 512x512 random uint8 image, `repeat=3`, `number=100000` w/ `timeit.repeat`, aka taking the minimum mean execution time (ms) across 3 runs w/ different random images w/ 100000 calls in each run

- expand only w/ garbage collection: 0.0015ms
- expand + recast w/ garbage collection: 0.59ms
- performance gain (recast/expand): 509x

- expand only without gc: 0.0016ms
- expand + recast without gc: 0.51ms
- performance gain (recast/expand): 442x

so the recasting takes a p long time, especially important for the live class where iterations should be in the ~milliseconds range to keep up with high-speed video. the extra garbage collection time alone is ~60x longer than the view-only method. You know what happens downstream better than I do, so could be that if it doesn't happen here it'll be converted to float later, but live module should have unidirectional image transformation pathway, (eg. if uint8 RGB is the desired form, once it becomes uint8 RGB it stays that way.)

testing code (python==3.6.9, numpy==1.17.4):
`
import numpy as np
import timeit

print('Comparing expand_dims w/ and w/o recasting')

n_repeats = 3
n_number  = 100000

img_shape = (512,512)

divider = '\n------------------------\n'

print('repeat = {}, number = {}'.format(n_repeats, n_number)+divider)

# setup strings
setup_str = 'import numpy as np; img = np.random.randint(0,255,({ht},{wd},3),dtype="uint8")'.format(ht=img_shape[0], wd=img_shape[1])
setup_str_gc = 'gc.enable(); ' + setup_str

# commands to compare. assigned to x because otherwise
test_str_expand = 'x = np.expand_dims(img, axis=0)'
test_str_recast = test_str_expand + '.astype(float)'


# just giving expand dims view
# re-enable garbage collection for both tests since
expand_only = timeit.repeat(test_str_expand,
              setup=setup_str_gc,
              repeat=n_repeats, number=n_number)

print('completed test 1')

# also doing recasting
expand_recast = timeit.repeat(test_str_recast,
              setup=setup_str_gc,
              repeat=n_repeats, number=n_number)

print('completed test 2')

# test without garbage collection
# also doing recasting
expand_only_nogc = timeit.repeat(test_str_expand,
              setup=setup_str,
              repeat=n_repeats, number=n_number)

print('completed test 3')

# also doing recasting
expand_recast_nogc = timeit.repeat(test_str_recast,
              setup=setup_str,
              repeat=n_repeats, number=n_number)

print('completed tests'+divider)

# divide by n_number to get time per operation
expand_only = np.divide(expand_only, n_number)
expand_recast = np.divide(expand_recast, n_number)
expand_only_nogc = np.divide(expand_only_nogc, n_number)
expand_recast_nogc = np.divide(expand_recast_nogc, n_number)

print("Tests with garbage collection\n")

print('Expand dims only: \nmin: {}\nall: {}\n'.format(np.min(expand_only), expand_only))
print('Expand and recast:\nmin: {}\nall: {}\n'.format(np.min(expand_recast), expand_recast))

speed_advantage = np.round(np.min(expand_recast)/np.min(expand_only), 2)
print('Speed advantage (expand/(expand+recast)): {}x faster'.format(speed_advantage))

print(divider)
print("Tests without garbage collection\n")

print('Expand dims only: \nmin: {}\nall: {}\n'.format(np.min(expand_only_nogc), expand_only_nogc))
print('Expand and recast:\nmin: {}\nall: {}\n'.format(np.min(expand_recast_nogc), expand_recast_nogc))

speed_advantage_nogc = np.round(np.min(expand_recast_nogc)/np.min(expand_only_nogc), 2)
print('Speed advantage (expand/(expand+recast)): {}x faster'.format(speed_advantage_nogc))

`

Reply:
effect scales w/ image size as one might expect (gc  overhead doesn't scale).

w/ a 1024x1024 img,
expand only w/ garbage collection: 0.0017ms
expand + recast w/ garbage collection: 2.565ms
performance gain (recast/expand): 2189x
expand only without gc: 0.0017ms
expand + recast without gc: 2.637ms 
performance gain (recast/expand): 2241x